# Rep 1 Python Script
#### Group Members
* Luis
* Diego
* Álvaro


In [105]:
pip install pyreadr==0.4.2

Note: you may need to restart the kernel to use updated packages.


In [106]:
#Import packages
import pandas as pd
import numpy as np
import pyreadr as rr 
import math

In [107]:
#import .Rdata 
rdata_read = rr.read_r(r"../../data/wage2015_subsample_inference.Rdata")
#select data
data = rdata_read[ 'data' ]
print(data.shape)
print(type(data))

(5150, 20)
<class 'pandas.core.frame.DataFrame'>


In [108]:
#check data
data

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we  \
rownames                                                                     
10         9.615385  2.263364  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
12        48.076923  3.872802  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
15        11.057692  2.403126  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
18        13.942308  2.634928  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
19        28.846154  3.361977  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
...             ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32620     14.769231  2.692546  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
32624     23.076923  3.138833  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0   
32626     38.461538  3.649659  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
32631     32.967033  3.495508  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32643     17.307692  2.851151  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   

           ne  exp1  exp2    exp3     exp4   occ occ2   ind ind2  
rownames                                                          
10        1.0   7.0  0.49   0.343   0.2401  3600   11  8370   18  
12        1.0  31.0  9.61  29.791  92.3521  3050   10  5070    9  
15        1.0  18.0  3.24   5.832  10.4976  6260   19   770    4  
18        1.0  25.0  6.25  15.625  39.0625   420    1  6990   12  
19        1.0  22.0  4.84  10.648  23.4256  2015    6  9470   22  
...       ...   ...   ...     ...      ...   ...  ...   ...  ...  
32620     0.0   9.0  0.81   0.729   0.6561  4700   16  4970    9  
32624     0.0  12.0  1.44   1.728   2.0736  4110   13  8680   20  
32626     0.0  11.0  1.21   1.331   1.4641  1550    4  3680    6  
32631     0.0  10.0  1.00   1.000   1.0000  2920    9  6570   11  
32643     0.0  14.0  1.96   2.744   3.8416  1610    5  7460   14  

[5150 rows x 20 columns]

#### Variable description
* occ : occupational classification
* ind : industry classification
* lwage : log hourly wage
* sex : gender (1 female) (0 male)
* shs : some high school
* hsg : High school graduated
* scl : Some College
* clg: College Graduate
* ad: Advanced Degree
* ne: Northeast
* mw: Midwest
* so: South
* we: West
* exp1: experience

In [109]:
data.info()
#No null data, it's ok!

<class 'pandas.core.frame.DataFrame'>
Index: 5150 entries, 10 to 32643
Data columns (total 20 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   wage    5150 non-null   float64 
 1   lwage   5150 non-null   float64 
 2   sex     5150 non-null   float64 
 3   shs     5150 non-null   float64 
 4   hsg     5150 non-null   float64 
 5   scl     5150 non-null   float64 
 6   clg     5150 non-null   float64 
 7   ad      5150 non-null   float64 
 8   mw      5150 non-null   float64 
 9   so      5150 non-null   float64 
 10  we      5150 non-null   float64 
 11  ne      5150 non-null   float64 
 12  exp1    5150 non-null   float64 
 13  exp2    5150 non-null   float64 
 14  exp3    5150 non-null   float64 
 15  exp4    5150 non-null   float64 
 16  occ     5150 non-null   category
 17  occ2    5150 non-null   category
 18  ind     5150 non-null   category
 19  ind2    5150 non-null   category
dtypes: category(4), float64(16)
memory usage: 736.3+ KB


In [110]:
#Summary statistics
data.describe()

wage        lwage          sex          shs          hsg  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean     23.410410     2.970787     0.444466     0.023301     0.243883   
std      21.003016     0.570385     0.496955     0.150872     0.429465   
min       3.021978     1.105912     0.000000     0.000000     0.000000   
25%      13.461538     2.599837     0.000000     0.000000     0.000000   
50%      19.230769     2.956512     0.000000     0.000000     0.000000   
75%      27.777778     3.324236     1.000000     0.000000     0.000000   
max     528.845673     6.270697     1.000000     1.000000     1.000000   

               scl          clg           ad           mw           so  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.278058     0.317670     0.137087     0.259612     0.296505   
std       0.448086     0.465616     0.343973     0.438464     0.456761   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     1.000000     0.000000     1.000000     1.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

                we           ne         exp1         exp2         exp3  \
count  5150.000000  5150.000000  5150.000000  5150.000000  5150.000000   
mean      0.216117     0.227767    13.760583     3.018925     8.235867   
std       0.411635     0.419432    10.609465     4.000904    14.488962   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     5.000000     0.250000     0.125000   
50%       0.000000     0.000000    10.000000     1.000000     1.000000   
75%       0.000000     0.000000    21.000000     4.410000     9.261000   
max       1.000000     1.000000    47.000000    22.090000   103.823000   

              exp4  
count  5150.000000  
mean     25.118038  
std      53.530225  
min       0.000000  
25%       0.062500  
50%       1.000000  
75%      19.448100  
max     487.968100

In [ ]:
#Let's focus on some college and complete college educated individuals

Z = data [ ["lwage","sex","scl","clg","ad","ne","mw","so","we","exp1"] ]

data_scl = data[data[ 'scl' ] == 1 ]
Z_scl = data_scl[ ["lwage","sex","scl","clg","ad","ne","mw","so","we","exp1"] ]

data_clg = data[ data[ 'clg' ] == 1 ]
Z_clg = data_clg[ [ "lwage","sex","scl","clg","ad","ne","mw","so","we","exp1" ] ]

table = np.zeros( (10, 3) )
table[:, 0] = Z.mean().values
table[:, 1] = Z_scl.mean().values
table[:, 2] = Z_clg.mean().values
table_pandas = pd.DataFrame( table, columns = [ 'All', 'Some College Ed.', 'College Ed.'])
table_pandas.index = ["Log Wage","Sex","Some College","Collage Graduate","Advanced Degree", "Northeast","Midwest","South","West","Experience"]

table_pandas


### Unconditional difference

In [111]:
# Difference in completed college v. some college workers

diff_mean= data_clg['lwage'].mean() - data_scl['lwage'].mean()

print("Comparing by means, being a completed college workers brings ", round(diff_mean*100, 2), "% more salary than a some college worker")

Comparing by means, being a completed college workers brings  28.63 % more salary than a some college worker


### Select our interested subset

In [112]:
#An unconditional comparision is generally biased. Let's make some regressions to apply some Econometrics to capture the causal effect of college education in wages.

#install pkgs
!pip install statsmodels

import statsmodels.api as sm
import statsmodels.formula.api as smf

In [113]:
#previously clean our dataset (some college v. college completed subset)
subset= data.copy()
subset

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so   we  \
rownames                                                                     
10         9.615385  2.263364  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
12        48.076923  3.872802  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
15        11.057692  2.403126  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   
18        13.942308  2.634928  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
19        28.846154  3.361977  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0   
...             ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32620     14.769231  2.692546  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0   
32624     23.076923  3.138833  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0   
32626     38.461538  3.649659  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
32631     32.967033  3.495508  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0   
32643     17.307692  2.851151  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   

           ne  exp1  exp2    exp3     exp4   occ occ2   ind ind2  
rownames                                                          
10        1.0   7.0  0.49   0.343   0.2401  3600   11  8370   18  
12        1.0  31.0  9.61  29.791  92.3521  3050   10  5070    9  
15        1.0  18.0  3.24   5.832  10.4976  6260   19   770    4  
18        1.0  25.0  6.25  15.625  39.0625   420    1  6990   12  
19        1.0  22.0  4.84  10.648  23.4256  2015    6  9470   22  
...       ...   ...   ...     ...      ...   ...  ...   ...  ...  
32620     0.0   9.0  0.81   0.729   0.6561  4700   16  4970    9  
32624     0.0  12.0  1.44   1.728   2.0736  4110   13  8680   20  
32626     0.0  11.0  1.21   1.331   1.4641  1550    4  3680    6  
32631     0.0  10.0  1.00   1.000   1.0000  2920    9  6570   11  
32643     0.0  14.0  1.96   2.744   3.8416  1610    5  7460   14  

[5150 rows x 20 columns]

In [114]:
subset['clg_comp'] = np.where(subset['clg']==1, 1, 0)

subset

wage     lwage  sex  shs  hsg  scl  clg   ad   mw   so  ...  \
rownames                                                               ...   
10         9.615385  2.263364  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...   
12        48.076923  3.872802  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...   
15        11.057692  2.403126  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...   
18        13.942308  2.634928  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...   
19        28.846154  3.361977  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...   
...             ...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
32620     14.769231  2.692546  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...   
32624     23.076923  3.138833  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  ...   
32626     38.461538  3.649659  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...   
32631     32.967033  3.495508  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...   
32643     17.307692  2.851151  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...   

           ne  exp1  exp2    exp3     exp4   occ occ2   ind ind2 clg_comp  
rownames                                                                   
10        1.0   7.0  0.49   0.343   0.2401  3600   11  8370   18        1  
12        1.0  31.0  9.61  29.791  92.3521  3050   10  5070    9        1  
15        1.0  18.0  3.24   5.832  10.4976  6260   19   770    4        0  
18        1.0  25.0  6.25  15.625  39.0625   420    1  6990   12        0  
19        1.0  22.0  4.84  10.648  23.4256  2015    6  9470   22        1  
...       ...   ...   ...     ...      ...   ...  ...   ...  ...      ...  
32620     0.0   9.0  0.81   0.729   0.6561  4700   16  4970    9        1  
32624     0.0  12.0  1.44   1.728   2.0736  4110   13  8680   20        0  
32626     0.0  11.0  1.21   1.331   1.4641  1550    4  3680    6        0  
32631     0.0  10.0  1.00   1.000   1.0000  2920    9  6570   11        0  
32643     0.0  14.0  1.96   2.744   3.8416  1610    5  7460   14        0  

[5150 rows x 21 columns]

In [115]:
#dispense with individual with no completed nor some college education
subset = subset[subset['shs']==0]
subset = subset[subset['hsg']==0]
subset = subset[subset['ad']==0]

#dispense with clg and scl columns
subset.drop(['shs','hsg','ad'], axis='columns', inplace=True)

subset

wage     lwage  sex  scl  clg   mw   so   we   ne  exp1  exp2  \
rownames                                                                       
10         9.615385  2.263364  1.0  0.0  1.0  0.0  0.0  0.0  1.0   7.0  0.49   
12        48.076923  3.872802  0.0  0.0  1.0  0.0  0.0  0.0  1.0  31.0  9.61   
19        28.846154  3.361977  1.0  0.0  1.0  0.0  0.0  0.0  1.0  22.0  4.84   
30        11.730769  2.462215  1.0  0.0  1.0  0.0  0.0  0.0  1.0   1.0  0.01   
71        19.230769  2.956512  1.0  0.0  1.0  0.0  0.0  0.0  1.0   4.0  0.16   
...             ...       ...  ...  ...  ...  ...  ...  ...  ...   ...   ...   
32596     45.546559  3.818735  1.0  0.0  1.0  0.0  0.0  1.0  0.0   5.0  0.25   
32606     24.038462  3.179655  1.0  0.0  1.0  0.0  0.0  1.0  0.0  17.0  2.89   
32619     13.846154  2.628007  0.0  0.0  1.0  0.0  0.0  1.0  0.0  10.0  1.00   
32620     14.769231  2.692546  0.0  0.0  1.0  0.0  0.0  1.0  0.0   9.0  0.81   
32624     23.076923  3.138833  1.0  1.0  0.0  0.0  0.0  1.0  0.0  12.0  1.44   

            exp3     exp4   occ occ2   ind ind2  clg_comp  
rownames                                                   
10         0.343   0.2401  3600   11  8370   18         1  
12        29.791  92.3521  3050   10  5070    9         1  
19        10.648  23.4256  2015    6  9470   22         1  
30         0.001   0.0001  1650    5  7460   14         1  
71         0.064   0.0256  3255   10  8190   18         1  
...          ...      ...   ...  ...   ...  ...       ...  
32596      0.125   0.0625  3255   10  8190   18         1  
32606      4.913   8.3521  2550    8  9480   22         1  
32619      1.000   1.0000   800    2   770    4         1  
32620      0.729   0.6561  4700   16  4970    9         1  
32624      1.728   2.0736  4110   13  8680   20         0  

[3068 rows x 18 columns]

### Make regressions to find the causal effect of college education in wages.

In [142]:
#remember: ~ = Alt + 126

nocontrol_model = smf.ols(formula = 'lwage ~ clg_comp', data=subset )
nocontrol_est = nocontrol_model.fit().summary2().tables[1]['Coef.']['clg_comp']
nocontrol_est
nocontrol_se2 = nocontrol_model.fit().summary2().tables[1]['Std.Err.']['clg_comp']

# robust standar erros
HCV_coefs = nocontrol_model.fit().cov_HC0
nocontrol_se = np.power( HCV_coefs.diagonal() , 0.5)[1]
nocontrol_se

# print unconditional effect of gender and the corresponding standard error

print( f'The estimated college coefficient is {nocontrol_est} and the corresponding standard error is {nocontrol_se2}' )
print( f'The estimated college coefficient is {nocontrol_est} and the corresponding robust standard error is {nocontrol_se}','\n' )

nocontrol_model.fit().conf_int( alpha=0.05 ).loc[['clg_comp']]

The estimated college coefficient is 0.2863352099445873 and the corresponding standard error is 0.01901742129477679
The estimated college coefficient is 0.2863352099445873 and the corresponding robust standard error is 0.019018119623101924 



0         1
clg_comp  0.249047  0.323623

We obtain the unconditional college wage gap $28,63 \%$ is the same as the no control OLS regression estimator, $0.2863$

Will it change if we make a regression using controls to get a better estimator for the causal effect of college education? \
Let's see it!

### OLS regression with controls

In [127]:
#Write the equation to run the regression
flex = 'lwage ~ clg_comp + sex + (exp1+exp2+exp3+exp4)*(occ2+ind2+mw+so+we)'

#Regression using multiple controls:
control_model = smf.ols( formula = flex, data = subset )
control_est = control_model.fit().summary2().tables[1]['Coef.']['clg_comp']
control_est

#estimators
print(control_model.fit().summary2().tables[1])

HCV_coefs = control_model.fit().cov_HC0
control_se = np.power( HCV_coefs.diagonal() , 0.5)[42]  # error standard for sex's coefficients 
control_se
print("")
#estimator of interest and std error
print( f"Coefficient for OLS with controls {control_est} and the corresponding robust standard error is {control_se}" )

# confidence interval
control_model.fit().conf_int( alpha=0.05 ).loc[['clg_comp']]


               Coef.  Std.Err.         t         P>|t|    [0.025    0.975]
Intercept   3.009600  0.318297  9.455324  6.511439e-21  2.385484  3.633716
occ2[T.10]  0.087324  0.242685  0.359825  7.190044e-01 -0.388533  0.563182
occ2[T.11] -0.509177  0.435109 -1.170229  2.420070e-01 -1.362337  0.343984
occ2[T.12]  0.188733  0.340697  0.553963  5.796480e-01 -0.479305  0.856771
occ2[T.13] -0.200005  0.271454 -0.736791  4.613106e-01 -0.732271  0.332262
...              ...       ...       ...           ...       ...       ...
exp3:so     0.006920  0.172433  0.040131  9.679916e-01 -0.331186  0.345026
exp3:we    -0.235683  0.183925 -1.281406  2.001556e-01 -0.596323  0.124957
exp4:mw     0.016045  0.025726  0.623660  5.329012e-01 -0.034400  0.066489
exp4:so    -0.002222  0.022498 -0.098772  9.213261e-01 -0.046336  0.041892
exp4:we     0.028822  0.023758  1.213148  2.251739e-01 -0.017763  0.075408

[227 rows x 6 columns]

Coefficient for OLS with controls 0.23686999754654117 and the corresponding

0         1
clg_comp  0.197619  0.276121

It seems that the estimator of college education was overstated (positive biased), because when controls are used, the estimator value reduces from $28,63 \%$ to $23,69\%$. The use of variables of controls such as sex, occupation and region, also known as worker characteristics, allow us to obtain a better estimator because is using fixed (and more) information to make an efficient estimation.

### Partialling-out using OLS

Make a partial regression of Y to X0 and X1 to X0, and then find the causal effect of $\beta_1$,using Frisch–Waugh–Lovell theorem.

In [128]:
#model for Y
flex_y = 'lwage ~  sex + (exp1+exp2+exp3+exp4)*(occ2+ind2+mw+so+we)'

#model for X1
flex_x = 'clg_comp ~   sex + (exp1+exp2+exp3+exp4)*(occ2+ind2+mw+so+we)'


In [131]:
#partialling-out the linear effect of W from Y
t_Y = smf.ols(formula = flex_y , data=subset).fit().resid

#partialling-out the linear effect of W from X1
t_X = smf.ols(formula = flex_x , data=subset).fit().resid

#join the residues 
data_res = pd.DataFrame(np.vstack((t_Y.values, t_X.values)).T, columns=['t_Y', 't_X'])

In [141]:
#regression
partial_fit = smf.ols(formula = 't_Y ~ t_X', data=data_res).fit()

#take the estimator and std. error
partial_est = partial_fit.summary2().tables[1]['Coef.']['t_X']

HCV_coefs = partial_fit.cov_HC0
partial_se=np.power(HCV_coefs.diagonal(), 0.5)[1]

print('')
print(f'Coefficient for X via partialling-out {partial_est} and the corresponding robust standard error is {partial_se}')
print('')

#confidence interval
partial_fit.conf_int( alpha=0.5).loc[['t_X']]



Coefficient for X via partialling-out 0.23686999754654472 and the corresponding robust standard error is 0.020252007635838558



0         1
t_X  0.223869  0.249871

Now, we used 4 methods to obtain the estimator for college education effect in wages (%): \
* Using difference of means, we obtain a coefficient of $28,63 \%$.
* Using OLS with no controls (regress Y~X), we obtain a coefficient of $28,63 \%$, in a confidence band between [0.249 - 0.3236] and a robust standard error of 0.0190.
* Using OLS controlled by fixed effects (worker charasteristics), we obtain a coefficient of $23,69 \%$, in a confidence band between [0.197619	- 0.276121] and a robust standard error of 0.0202. Here, the fixed effects estimator is lower than the previous one, what means that the previous estimations were overstated
* Using partially-out OLS (using Frisch–Waugh–Lovell theorem), we obtain a coefficient of $23,68 \%$, in a confidence band between [0.223869 - 0.249871] and a robust standard error of 0.02025. This estimator has the smallest conficende distance, and it is very similar to the fixed effect estimator. Possibly, this may be a BLUE.